psutil/arch/windows/process_info.c

```diff
@@ -804,10 +870,72 @@ psutil_get_cmdline(long pid) {
     PyObject *py_unicode = NULL;
     LPWSTR *szArglist = NULL;
     int nArgs, i;
+    HANDLE hProcess;
+    int windows_version;
+    ULONG ret_length = 4096;
+    NTSTATUS status;
+    char * cmdline_buffer = NULL;
+    WCHAR * cmdline_buffer_wchar = NULL;
+    PUNICODE_STRING tmp = NULL;
+    DWORD string_size;
+    _NtQueryInformationProcess NtQueryInformationProcess = NULL;
+
+    windows_version = get_windows_version();
+    if (windows_version >= WINDOWS_81) {
+        // by defaut, still use PEB (if command line params have been patched in PEB, we will get the actual ones)
+        if (psutil_get_process_data(pid, KIND_CMDLINE, &data, &size) != 0) {
```

Comment:
Mmmm... I'm a bit confused about this. It appears regardless of the Windows version you first attempt to use the PEB method first, and fallback on using `NtQueryInformationProcess` in case of failure on Win >= 8.1. Wouldn't it make sense to always use `NtQueryInformationProcess` on Win >= 8.1? I a bit against trying 2 different methods and suppressing error in case the first one fails. Unless there's a good reason I think it's better to always use 1 method only.

Reply:
Reading the PEB to get the command line parameters still seem to be the best method if somebody has tampered with the parameters after creating the process.
For instance, create a process as suspended, patch the command line in its PEB and unfreeze it.
The process will use the "new" parameters whereas the system (with `NtQueryInformationProcess`) will give you the "old" ones (see here : https://blog.xpnsec.com/how-to-argue-like-cobalt-strike/)
By default, reading from PEB is still allowed if one has sufficient privileges except for protected processes where one must use `NtQueryInformationProcess`.
So the PEB method will work for almost all processes on windows 8.1+ except the few (4-5) protected processes lying around (where we can fallback on NtQIP), and it will accurately report the actual command line parameters used

Reply:
OK, I see. In this case I would recommend the following:
* use `psutil_handle_from_pid(pid, PROCESS_QUERY_INFORMATION | PROCESS_VM_READ)`
* if it succeeds use the PEB method (and only that)
* if it doesn't (for any reason), use `NtQueryInformationProcess` if Win >= 8.1, else fail
In both cases you should simply close the returned handle: the only reason to get the handle is to see if you have permissions to do so. 
Makes sense?

Also, please add a full explanation of why one method is preferred over the other one (you can include the information you wrote in the comment above).

Reply:
Actually, I think using `psutil_handle_from_pid(pid, PROCESS_QUERY_INFORMATION | PROCESS_VM_READ)` is redundant as reading from PEB will also open process with those rights, so for every process, we would open/close it twice

Reply:
As it is, the code works. What I could do is add an explanation as why the PEB method is used first for windows >= 8.1 as you propose, but go with this code anyway, no?

Reply:
Yes but as-is it tries the other method for any error. I would like to limit the check to ERROR_ACCESS_DENIED. If it fails for any other reason I think it's saner to just raise exception.

Reply:
So something like this (not tested):

```C
    int ret;

    ret = psutil_get_process_data(pid, KIND_CWD, &data, &size);
    if ((ret != 0) && 
            (GetLastError() == ERROR_ACCESS_DENIED) && 
            (windows_version >= WINDOWS_81)) 
    {
        // try NtQueryInformationProcess()
        PyErr_Clear();
        ...
    }
    else {
        goto out;
    }
```

Reply:
...also, I suggest to put the whole `NtQueryInformationProcess()` in a separate function, similarly to what it has been done for `psutil_get_process_data`

Reply:
hint: function name could be `psutil_get_cmdline_data`